File: src/main/java/no/fintlabs/validation/constraints/InstanceValueKeysAreDefinedInMetadata.java
package no.fintlabs.validation.constraints;

import javax.validation.Constraint;
import javax.validation.Payload;
import java.lang.annotation.Retention;

import static java.lang.annotation.RetentionPolicy.RUNTIME;

@Retention(RUNTIME)
@Constraint(validatedBy = InstanceValueKeysAreDefinedInMetadataValidator.class)
public @interface InstanceValueKeysAreDefinedInMetadata {

    String KEYS_MISSING_IN_METADATA = "keysMissingInMetadata";

    String message() default "contains references to instance values that are not defined in the metadata: {" + KEYS_MISSING_IN_METADATA + "}";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};

}

File: src/main/kotlin/no/novari/flyt/configuration/validation/constraints/InstanceValueKeysAreDefinedInMetadata.kt
package no.novari.flyt.configuration.validation.constraints

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.reflect.KClass

@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.CLASS)
@Constraint(validatedBy = [InstanceValueKeysAreDefinedInMetadataValidator::class])
annotation class InstanceValueKeysAreDefinedInMetadata(
    val message: String =
        "contains references to instance values that are not defined in the metadata: {$KEYS_MISSING_IN_METADATA}",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = [],
) {
    companion object {
        const val KEYS_MISSING_IN_METADATA = "keysMissingInMetadata"
    }
}
